File: src/core/ipc/blockstore.ts
/**
 * Block store IPC wrappers for the Persistent Window Architecture.
 *
 * All commands invoke the Rust backend via Tauri. These functions are
 * Tauri-only — callers must guard with `isTauriContext()` before use.
 *
 * @module ipc/blockstore
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  BlockContent,
  BlockMeta,
  BlockSearchMatch,
  DocumentHandle,
  WindowUpdateResult,
} from '@/types/blockstore';

// Re-export types callers commonly need alongside these wrappers
export type { DocumentHandle, BlockContent, BlockMeta, BlockSearchMatch, WindowUpdateResult };

// ── Document lifecycle ─────────────────────────────────────────────────────────

/**
 * Open a document: read from disk, scan into blocks, cache in the Rust backend.
 * Returns full block metadata (no content) for scrollbar initialisation.
 */
export async function openDocument(path: string): Promise<DocumentHandle> {
  console.log('[INFO] [IPC/blockstore] openDocument:', path);
  return invoke<DocumentHandle>('open_document', { path });
}

/**
 * Remove a document from the Rust in-memory store.
 * Call before switching documents or switching to source mode.
 */
export async function closeDocument(docId: string): Promise<void> {
  console.log('[INFO] [IPC/blockstore] closeDocument:', docId);
  return invoke<void>('close_document', { docId });
}

// ── Block reads ────────────────────────────────────────────────────────────────

/**
 * Fetch markdown content for blocks `[start, end)`.
 * Typical viewport window: ~200 blocks with a ±20 block buffer.
 */
export async function getBlocks(
  docId: string,
  start: number,
  end: number,
): Promise<BlockContent[]> {
  return invoke<BlockContent[]>('get_blocks', { docId, start, end });
}

// ── Block writes ───────────────────────────────────────────────────────────────

/**
 * Replace the visible block range with a re-scan of `windowMarkdown`.
 *
 * Handles block splits (user added a blank line) and merges (user deleted a
 * blank line) transparently. Returns updated full block metadata so the
 * frontend can recalibrate the synthetic scrollbar.
 */
export async function updateVisibleWindow(
  docId: string,
  startBlock: number,
  endBlock: number,
  windowMarkdown: string,
): Promise<WindowUpdateResult> {
  console.log('[INFO] [IPC/blockstore] updateVisibleWindow:', docId, startBlock, endBlock);
  return invoke<WindowUpdateResult>('update_visible_window', {
    docId,
    startBlock,
    endBlock,
    windowMarkdown,
  });
}

/**
 * Refine the estimated render height of a single block with a DOM measurement.
 * Called by PersistentWindow's ResizeObserver after a block is rendered.
 */
export async function updateBlockHeight(
  docId: string,
  blockId: number,
  height: number,
): Promise<void> {
  return invoke<void>('update_block_height', { docId, blockId, height });
}

// ── Save ───────────────────────────────────────────────────────────────────────

/**
 * Reassemble all blocks and write the document to disk.
 * Skips the write if the document is not dirty.
 */
export async function saveDocument(docId: string): Promise<void> {
  console.log('[INFO] [IPC/blockstore] saveDocument:', docId);
  return invoke<void>('save_document', { docId });
}

// ── Search ─────────────────────────────────────────────────────────────────────

/**
 * Case-insensitive search across all blocks in a document.
 * Returns one match per block (first occurrence only).
 */
export async function searchBlocks(
  docId: string,
  query: string,
): Promise<BlockSearchMatch[]> {
  console.log('[INFO] [IPC/blockstore] searchBlocks:', docId, query);
  return invoke<BlockSearchMatch[]>('search_blocks', { docId, query });
}

File: src-tauri/src/commands/blockstore.rs
//! Block store IPC commands for the Persistent Window Architecture.
//!
//! All commands take `State<'_, DocumentStore>` and delegate to the
//! [`DocumentStore`] methods. Return types are serialised as JSON over IPC.

use tauri::State;
use tracing::info;

use crate::blockstore::{
    BlockContent, BlockSearchMatch, BlockUpdate, DocumentHandle, DocumentStore,
    WindowUpdateResult,
};

/// Open a document: read from disk, scan into blocks, cache in memory.
///
/// Returns full block metadata (no content) so the frontend can initialise
/// the synthetic scrollbar and ViewportCoordinator.
#[tauri::command]
pub async fn open_document(
    path: String,
    store: State<'_, DocumentStore>,
) -> Result<DocumentHandle, String> {
    info!("[INFO] [blockstore_cmd] open_document: {}", path);
    store.open_document(&path)
}

/// Return markdown content for blocks `[start, end)`.
#[tauri::command]
pub async fn get_blocks(
    doc_id: String,
    start: usize,
    end: usize,
    store: State<'_, DocumentStore>,
) -> Result<Vec<BlockContent>, String> {
    info!(
        "[INFO] [blockstore_cmd] get_blocks: {} [{}, {})",
        doc_id, start, end
    );
    store.get_blocks(&doc_id, start, end)
}

/// Apply individual block updates (content-only, no count change).
#[tauri::command]
pub async fn update_blocks(
    doc_id: String,
    updates: Vec<BlockUpdate>,
    store: State<'_, DocumentStore>,
) -> Result<(), String> {
    info!(
        "[INFO] [blockstore_cmd] update_blocks: {} ({} updates)",
        doc_id,
        updates.len()
    );
    store.update_blocks(&doc_id, updates)
}

/// Replace the visible block range with a re-scan of `window_markdown`.
///
/// Handles block splits (user added blank line) and merges (user deleted
/// blank line) by splicing the in-memory block array.
///
/// Returns updated full block metadata so the frontend can recalibrate the
/// synthetic scrollbar.
#[tauri::command]
pub async fn update_visible_window(
    doc_id: String,
    start_block: usize,
    end_block: usize,
    window_markdown: String,
    store: State<'_, DocumentStore>,
) -> Result<WindowUpdateResult, String> {
    info!(
        "[INFO] [blockstore_cmd] update_visible_window: {} [{}, {})",
        doc_id, start_block, end_block
    );
    store.update_visible_window(&doc_id, start_block, end_block, &window_markdown)
}

/// Reassemble all blocks and write the document to disk.
#[tauri::command]
pub async fn save_document(
    doc_id: String,
    store: State<'_, DocumentStore>,
) -> Result<(), String> {
    info!("[INFO] [blockstore_cmd] save_document: {}", doc_id);
    store.save_document(&doc_id)
}

/// Remove a document from the in-memory store (called on document close or
/// mode switch to source view).
#[tauri::command]
pub async fn close_document(
    doc_id: String,
    store: State<'_, DocumentStore>,
) -> Result<(), String> {
    info!("[INFO] [blockstore_cmd] close_document: {}", doc_id);
    store.close_document(&doc_id)
}

/// Case-insensitive search across all blocks in a document.
#[tauri::command]
pub async fn search_blocks(
    doc_id: String,
    query: String,
    store: State<'_, DocumentStore>,
) -> Result<Vec<BlockSearchMatch>, String> {
    info!(
        "[INFO] [blockstore_cmd] search_blocks: {} query='{}'",
        doc_id, query
    );
    store.search_blocks(&doc_id, &query)
}
